heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/CacheCore.java

```diff
@@ -0,0 +1,543 @@
+        addException(exceptionLog);
+      }
+    }
+  }
+
+  /**
+   * make bucket id
+   *
+   * @param hi index of [component-instance]
+   * @param lo index of metric name
+   * @return bucket id
+   */
+  private long makeBucketId(int hi, int lo) {
+    return (((long) hi) << 32) | (lo & 0xffffffffL);
+  }
+
+  private void addMetric(TopologyMaster.MetricDatum metricDatum) {
+    String componentName = metricDatum.getComponentName();
+    String instanceId = metricDatum.getInstanceId();
+    String metricName = metricDatum.getName();
+    LOG.info(componentName + " " + instanceId + " " + metricName);
+    assureComponentInstance(componentName, instanceId);
+    assureMetricName(metricName);
+    // calc bucket idx
+    int idx1 = idxComponentInstance.get(componentName).get(instanceId);
+    int idx2 = idxMetricName.get(metricName);
+    long bucketId = makeBucketId(idx1, idx2);
+    LOG.info(Long.toHexString(bucketId) + " = "
+        + Integer.toHexString(idx1) + " " + Integer.toHexString(idx2));
+    // fetch the bucket
+    Map.Entry<Long, Map<Long, LinkedList<MetricDatapoint>>> entry =
+        cacheMetric.floorEntry(metricDatum.getTimestamp());
+    if (entry != null) {
+      Map<Long, LinkedList<MetricDatapoint>> locator = entry.getValue();
+      if (!locator.containsKey(bucketId)) {
+        locator.put(bucketId, new LinkedList<MetricDatapoint>());
+      }
+      LinkedList<MetricDatapoint> bucket = locator.get(bucketId);
+      // store the metric
+      MetricDatapoint datum = new MetricDatapoint();
+      datum.timestamp = metricDatum.getTimestamp();
+      datum.value = metricDatum.getValue();
+      bucket.offerFirst(datum);
+      LOG.info("addMetrics " + datum);
+    } else {
+      LOG.warning("too old metric: " + metricDatum);
+    }
+  }
+
+  private void addException(TopologyMaster.TmasterExceptionLog exceptionLog) {
+    String componentName = exceptionLog.getComponentName();
+    String instanceId = exceptionLog.getInstanceId();
+    assureComponentInstance(componentName, instanceId);
+    // get exception idx
+    int idx = idxComponentInstance.get(componentName).get(instanceId);
+    // fetch the bucket
+    if (!cacheException.containsKey(idx)) {
+      cacheException.put(idx, new LinkedList<ExceptionDatapoint>());
+    }
+    LinkedList<ExceptionDatapoint> bucket = cacheException.get(idx);
+    // store the exception
+    ExceptionDatapoint e = new ExceptionDatapoint();
+    e.componentName = exceptionLog.getComponentName();
+    e.hostname = exceptionLog.getHostname();
+    e.instanceId = exceptionLog.getInstanceId();
+    e.stackTrace = exceptionLog.getStacktrace();
+    e.lastTime = exceptionLog.getLasttime();
+    e.firstTime = exceptionLog.getFirsttime();
+    e.count = exceptionLog.getCount();
+    e.logging = exceptionLog.getLogging();
+    bucket.offerFirst(e);
+    // purge
+    while (bucket.size() > maxExceptionCount) {
+      LOG.warning("too many exception: " + exceptionLog);
```

Comment:
too many exception, so what? please provide more context to the reader of the logs so they know what this means. Does this mean we're dropping data? is there a bug, etc?